src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let clones = [state];
+
+  actions.forEach(({ type, extraData, keysToRemove }) => {
+    const lastState = clones[clones.length - 1];
+
+    switch (type) {
+      case 'addProperties':
+        clones = [...clones, {
+          ...lastState,
+          ...extraData,
+        }];
+        break;
+
+      case 'removeProperties':
+        const newState = { ...lastState };
+
+        keysToRemove.forEach(key => {
+          delete newState[key];
+        });
+
+        clones = [
+          ...clones,
+          newState,
+        ];
```

Comment:
Again, a lot of reassigning, we could it with the `push`
Also, you could move this code from `switch` instruction